doc/design/listener.md

```diff
@@ -0,0 +1,316 @@
+wait on events. So the contract is reflected in the design.
+
+- **Problem:** Since the Listener reacts on events and not states it requires
+    the knowledge by whom it was notified.
+- **Solution:**
+  - Every TriggerHandle has a unique id which is used as index in `ConditionNotifier`.
+  - When `ConditionNotifier::notify` is called the Listener is informed via the
+    `NotificationVector_t` return value from `ConditionListener::wait()` which
+    index notified him. Hence it is the same as the unique id of the TriggerHandle
+    the Listener knows which Triggerable notified him.
+
+#### Event_t
+
+##### Concurrency
+
+The Listener must be able to attach, detach events concurrently. Additionally,
+it supports that a callback can attach or detach further events or detach its
+corresponding event concurrently. Furthermore, the Listener supports that a
+callback is called concurrently while events are being attached/detached.
+
+To realize this we created the `Event_t` abstraction which is stored in an
+array called `m_events`. If we would like to attach or detach an event we either
+initialize `Event_t::init()` or reset `Event_t::reset()` the corresponding entry
+in the `m_events` array. The array has the advantage that the data structure
+itself never changes during runtime therefore it does not have to be thread-safe.
+
+The thread-safety must then be ensured by the `Event_t` class itself. Since
+every concurrent action is contained in `Event_t` we can use
+`concurrent::smart_lock` in combination with a `std::recursive_mutex` to
+guarantee the thread-safe access.
+
+1. Concurrent attach/detach event and callback execution, ensured by
+   thread-safe `Event_t`.
+2. Detaching itself from within a callback is ensured via the
+   `std::recursive_mutex`.
+3. Attaching/detaching arbitrary events from within a callback is ensured by
+   securing every `Event_t` object in `m_events` with a `concurrent::smart_lock`.
+   This would not be possible if the data structure itself had to be thread-safe.
+
+##### Lifetime
+
+Since the event contains everything which is required to handle events it has
+the responsibility to ensure the life-time of the TriggerHandle. This is done
+by the `m_invalidationCallback` which is called in `Event_t::reset()` to
+invalidate the TriggerHandle in the corresponding Triggerable. This is either
+done when an event is detached or the Listener goes out of scope.
+
+#### Triggerable (e.g. Subscriber)
+
+The Triggerable is a set of classes of which events can be attached to the Listener.
+
+It is possible to either attach a specific event of a class to the Listener or
+attach the class without providing an event.
+
+The basic idea is that the Listener creates a TriggerHandle whenever an event
+is attached and provides that TriggerHandle to the corresponding Triggerable.
+The Triggerable uses then the TriggerHandle to notify the Listener about events.
+
+##### Multiple Events
```

Comment:
Please consider switching multiple events with single events to have the simple case first